File: Data Structures & Algorithms/mergeSort/submission-1.java
// Definition for a pair.
// class Pair {
//     public int key;
//     public String value;
//
//     public Pair(int key, String value) {
//         this.key = key;
//         this.value = value;
//     }
// }
class Solution {
    public List<Pair> mergeSort(List<Pair> pairs) {
        // todo(): use .get().keys
        if (pairs.isEmpty()) return new ArrayList<>();
        return mergeIt(pairs, 0, pairs.size() - 1);
    }

    public List<Pair> mergeIt(List<Pair> pairs, int start, int end) {
        if (start - end == 0) return List.of(pairs.get(start));

        int mid = ((end - start) / 2) + start;
        var leftList = mergeIt(pairs, start, mid);
        var rightList = mergeIt(pairs, mid+1, end);

        List<Pair> res = new ArrayList<>();
        int l = 0;
        int r = 0;

        while (l < leftList.size() && r < rightList.size()) {
            if (leftList.get(l).key <= rightList.get(r).key) {
                res.add(leftList.get(l++));
            } else {
                res.add(rightList.get(r++));
            }
        }

        if (r >= rightList.size()) res.addAll(leftList.subList(l, leftList.size()));
        else res.addAll(rightList.subList(r, rightList.size()));

        return res;
    }
}

File: Data Structures & Algorithms/quickSort/submission-1.go
// Definition for a pair.
// type Pair struct {
//     Key   int
//     Value string
// }

type Solution struct {

}

func NewSolution() *Solution {
    return &Solution{}
}

func QuickSort(pairs []Pair) []Pair {
    return sortHelper(pairs, 0, len(pairs) - 1)
}

func sortHelper(pairs []Pair, start, end int) []Pair {
    if end <= start {
        return pairs
    }

    pvtIndx := end
    left := start
    for i:= start ;i < pvtIndx; i++ {
        fmt.Println("ireach")
        if pairs[i].Key < pairs[pvtIndx].Key {
            pairs[left], pairs[i] = pairs[i], pairs[left]
            left += 1
        }
    }
    pairs[left], pairs[pvtIndx] = pairs[pvtIndx], pairs[left]

    sortHelper(pairs, start, left-1)
    sortHelper(pairs, left + 1, end)
    return pairs
}
